File: src/board.rs
//! Board representation and helpers.

use crate::constants::{SQUARES, STARTING_POSITION};

pub type Board = [[char; 8]; 8];

/// Create a fresh board with the standard starting position.
pub fn new_board() -> Board {
    STARTING_POSITION
}

/// True if (col, row) is within 0..8.
#[inline]
pub fn in_bounds(col: i32, row: i32) -> bool {
    (0..SQUARES).contains(&col) && (0..SQUARES).contains(&row)
}

/// True if the square is empty.
#[inline]
pub fn is_empty(board: &Board, col: i32, row: i32) -> bool {
    board[row as usize][col as usize] == ' '
}

/// True if the piece at (col, row) is white (lowercase).
#[inline]
pub fn is_white_piece(c: char) -> bool {
    c.is_lowercase() && c != ' '
}

/// True if the piece at (col, row) is black (uppercase).
#[inline]
pub fn is_black_piece(c: char) -> bool {
    c.is_uppercase() && c != ' '
}

/// True if the piece at (col, row) belongs to the given side. white = true means white to move.
pub fn piece_is_side(board: &Board, col: i32, row: i32, white: bool) -> bool {
    let c = board[row as usize][col as usize];
    if white {
        is_white_piece(c)
    } else {
        is_black_piece(c)
    }
}

/// Get piece at (col, row). Returns space if out of bounds.
pub fn get_piece(board: &Board, col: i32, row: i32) -> char {
    if !in_bounds(col, row) {
        return ' ';
    }
    board[row as usize][col as usize]
}

/// Set piece at (col, row). No-op if out of bounds.
pub fn set_piece(board: &mut Board, col: i32, row: i32, piece: char) {
    if in_bounds(col, row) {
        board[row as usize][col as usize] = piece;
    }
}

/// Find the king of the given color. Returns (col, row) or None if not found.
pub fn find_king(board: &Board, white: bool) -> Option<(i32, i32)> {
    let king = if white { 'k' } else { 'K' };
    for r in 0..SQUARES {
        for c in 0..SQUARES {
            if board[r as usize][c as usize] == king {
                return Some((c, r));
            }
        }
    }
    None
}

File: src/board.py
"""Board representation and helpers."""

from .constants import SQUARES, STARTING_POSITION


def new_board():
    """Return a fresh board with the standard starting position (copy)."""
    return [row[:] for row in STARTING_POSITION]


def in_bounds(col, row):
    return 0 <= col < SQUARES and 0 <= row < SQUARES


def is_empty(board, col, row):
    return board[row][col] == ' '


def is_white_piece(c):
    return c.islower() and c != ' '


def is_black_piece(c):
    return c.isupper() and c != ' '


def piece_is_side(board, col, row, white):
    c = board[row][col]
    return is_white_piece(c) if white else is_black_piece(c)


def get_piece(board, col, row):
    if not in_bounds(col, row):
        return ' '
    return board[row][col]


def set_piece(board, col, row, piece):
    if in_bounds(col, row):
        board[row][col] = piece


def find_king(board, white):
    king = 'k' if white else 'K'
    for r in range(SQUARES):
        for c in range(SQUARES):
            if board[r][c] == king:
                return (c, r)
    return None
